### PR TITLE
Allow multiple relationships in ConnectorType

### PR DESCRIPTION
**Pull request by LucGenetier**
Allow multiple relationships in ConnectorType
Fix CompileTimeTypeWrapperRecordValue.TryGetField
Add Equals to ConnectorType

This update doesn't update CdpRecordType yet
DV and NL repos are not impacted by these changes

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs [hunk 1]
```diff
@@ -112,11 +112,7 @@ public class ConnectorType : SupportsConnectorErrors
         internal ISwaggerSchema Schema { get; private set; } = null;
 
         // Relationships to external tables
-        internal List<string> ExternalTables { get; set; }
-
-        internal string RelationshipName { get; set; }
```
Comment (LucGenetier) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:117]: >RelationshipName [](http://example.com/codeflow?start=24&length=16)

Removing as this is never used anywhere

File: src/libraries/Microsoft.PowerFx.Connectors/Tabular/Capabilities/ComplexColumnCapabilities.cs [hunk 2]
```diff
@@ -25,5 +25,10 @@ public void AddColumnCapability(string name, ColumnCapabilitiesBase capability)
 
             _childColumnsCapabilities.Add(name, capability);
         }
+
+        public override int GetHashCode()
+        {
+            throw new System.NotImplementedException();
```
Comment (LucGenetier) [inline src/libraries/Microsoft.PowerFx.Connectors/Tabular/Capabilities/ComplexColumnCapabilities.cs:31]: >NotImplementedException [](http://example.com/codeflow?start=29&length=23)

This class is coming from PowerApps and currently not used

File: src/libraries/Microsoft.PowerFx.Core/Public/Values/CompileTimeTypeWrapperRecordValue.cs [hunk 3]
```diff
@@ -44,10 +44,10 @@ public override bool TryShallowCopy(out FormulaValue copy)
         protected override bool TryGetField(FormulaType fieldType, string fieldName, out FormulaValue result)
         {
             if (Type.TryGetFieldType(fieldName, out var compileTimeType))
-            {
+            {                                
                 // Only return field which were specified via the expectedType (IE RecordType),
                 // because inner record value may have more fields than the expected type.
-                if (compileTimeType == fieldType && _fields.TryGetValue(fieldName, out result))
+                if (compileTimeType.Equals(fieldType) && _fields.TryGetValue(fieldName, out result))
```
Comment (LucGenetier) [inline src/libraries/Microsoft.PowerFx.Core/Public/Values/CompileTimeTypeWrapperRecordValue.cs:50]: >Equals [](http://example.com/codeflow?start=36&length=6)

Key change

File: src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs [hunk 4]
```diff
@@ -675,17 +675,16 @@ internal static ConnectorType GetConnectorType(this ISwaggerParameter openApiPar
...
+                            
                             ConnectorType propertyType = new SwaggerParameter(propLogicalName, schema.Required.Contains(propLogicalName), kv.Value, kv.Value.Extensions).GetConnectorType(settings.Stack(schemaIdentifier));
 
                             if (settings.SqlRelationships != null)
                             {
-                                SqlRelationship relationship = settings.SqlRelationships.FirstOrDefault(sr => sr.ColumnName == propLogicalName);
+                                IEnumerable<SqlRelationship> relationships = settings.SqlRelationships.Where(sr => sr.ColumnName == propLogicalName);
 
-                                if (relationship != null)
+                                if (relationships.Any())
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs:685]: >relationships.Any() [](http://example.com/codeflow?start=36&length=19)

Should this be? relationships.?Any() == true

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs [hunk 5]
```diff
@@ -148,9 +144,7 @@ internal ConnectorType(ISwaggerSchema schema, ISwaggerParameter openApiParameter
                 // SalesForce only
                 if (schema.ReferenceTo != null && schema.ReferenceTo.Count == 1)
                 {
-                    ExternalTables = new List<string>(schema.ReferenceTo);
-                    RelationshipName = schema.RelationshipName;
-                    ForeignKey = null; // SalesForce doesn't provide it, defaults to "Id"
+                    ExternalTables = new List<(string, string)>(schema.ReferenceTo.Select(r => (r, null as string /* Seems to always be Id */)));
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:147]: >null as string /* Seems to always be Id */ [](http://example.com/codeflow?start=99&length=42)

If its always Id, should this change from null -> "Id"?

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs [hunk 6]
```diff
@@ -336,5 +332,88 @@ private Dictionary<string, FormulaValue> GetEnum()
 
             return enumDisplayNames.Zip(enumValues, (dn, ev) => new KeyValuePair<string, FormulaValue>(dn, ev)).ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
         }
+
+        public bool Equals(ConnectorType other)
+        {
+            return this.Name == other.Name &&
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:338]: >this.Name == other.Name && [](http://example.com/codeflow?start=19&length=26)

We should check for null first.

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs [hunk 7]
```diff
@@ -20,7 +20,7 @@ namespace Microsoft.PowerFx.Connectors
     // FormulaType is used to represent the type of the parameter in the Power Fx expression as used in Power Apps
     // ConnectorType contains more details information coming from the swagger file and extensions
     [DebuggerDisplay("{FormulaType._type}")]
-    public class ConnectorType : SupportsConnectorErrors
+    public class ConnectorType : SupportsConnectorErrors, IEquatable<ConnectorType>
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:23]: >IEquatable<ConnectorType> [](http://example.com/codeflow?start=58&length=25)

Why do we need this?

File: src/libraries/Microsoft.PowerFx.Connectors/Tabular/Capabilities/ColumnCapabilitiesDefinition.cs [hunk 8]
```diff
@@ -33,5 +35,32 @@ public ColumnCapabilitiesDefinition(IEnumerable<string> filterFunction)
             // ex: lt, le, eq, ne, gt, ge, and, or, not, contains, startswith, endswith, countdistinct, day, month, year, time
             FilterFunctions = filterFunction;           
         }
+
+        public bool Equals(ColumnCapabilitiesDefinition other)
+        {
+            return Enumerable.SequenceEqual(this.FilterFunctions ?? Array.Empty<string>(), other.FilterFunctions ?? Array.Empty<string>()) &&
+                   this.QueryAlias == other.QueryAlias &&
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Connectors/Tabular/Capabilities/ColumnCapabilitiesDefinition.cs:42]: nit: check for null first

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs [hunk 9]
```diff
@@ -112,11 +112,7 @@ public class ConnectorType : SupportsConnectorErrors
         internal ISwaggerSchema Schema { get; private set; } = null;
 
         // Relationships to external tables
-        internal List<string> ExternalTables { get; set; }
-
-        internal string RelationshipName { get; set; }
-
-        internal string ForeignKey { get; set; }
+        internal List<(string foreignTable, string foreignKey)> ExternalTables { get; set; }
```
[the remark this answers is not in the document]
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:115]: on internal objects I don't think IEnumrable<> is necessary but let's change that

On the POCO, I agree it's a very good idea as this new class might evolve over time

File: src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs (hunk 4 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs:685]: Good optimization

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs (hunk 5 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:147]: The problem is that nobody has been able to confirm if it's "Id" or the primary key (and if the primary key is always "Id" in SalesForce) or if it could be something else.
I have also searched the SalesForce dev docs w/o finding anything.

As a result I prefer keeping "null" to say "I don't know" rather than asserting, maybe wrongly", "it"s Id"

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs (hunk 7 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:23]: IEquatable<T> is the interface implementing Equals(T)
It's a good practice to use this interface when you implement Equals

See [https://learn.microsoft.com/en-us/dotnet/api/system.iequatable-1?view=net-8.0](https://learn.microsoft.com/en-us/dotnet/api/system.iequatable-1?view=net-8.0)

The IEquatable<T> interface is used by generic collection objects such as Dictionary<TKey,TValue>, List<T>, and LinkedList<T> when testing for equality in such methods as Contains, IndexOf, LastIndexOf, and Remove. It should be implemented for any object that might be stored in a generic collection.

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.CdpExtensions.TryGetFieldExternalTableName(Microsoft.PowerFx.Types.RecordType,System.String,System.String@,System.String@)

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorRelationship.cs [hunk 10]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using Microsoft.PowerFx.Core.Utils;
+
+namespace Microsoft.PowerFx.Connectors
+{
+    public class ConnectorRelationship
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorRelationship.cs:8]: >ConnectorRelationship [](http://example.com/codeflow?start=17&length=21)

Is this CDP specific? 

If so, how do non-CDP shims (ServiceNow) leverage it?

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs (hunk 7 above)
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:23]: but why are we putting these in a collection? My concern is the complexity of .equals below. Is there some simplification we can apply here?   Ie, is it sufficient to just compare logical names; or do we need a full comparison?

Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:23]: We can probably simplify it a bit (logical names aren't enough) and i'll check
We need it due to the change in CompileTimeTypeWrapperRecordValue
IEquatable<T> is a best practice anyway, regardless of using collections

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs [hunk 11]
```diff
@@ -336,5 +342,92 @@ private Dictionary<string, FormulaValue> GetEnum()
...
+            if (other == null)
+            {
+                return false;
+            }
+
+            return this.Name == other.Name &&
+                   this.DisplayName == other.DisplayName &&
+                   this.Description == other.Description &&
+                   this.IsRequired == other.IsRequired &&
+                   Enumerable.SequenceEqual((IList<ConnectorType>)this.Fields ?? Array.Empty<ConnectorType>(), (IList<ConnectorType>)other.Fields ?? Array.Empty<ConnectorType>()) &&
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:357]: >(IList<ConnectorType>)this.Fields [](http://example.com/codeflow?start=44&length=33)

This is recursive - how do we verify this is not a stack overflow?

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorRelationship.cs (hunk 10 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorRelationship.cs:8]: for now, yes
as discussed this is a temporary thing
I'll add [Obsolete] to it

File: src/libraries/Microsoft.PowerFx.Connectors/Public/CdpExtensions.cs [hunk 12]
```diff
@@ -1,22 +1,23 @@
...
+using System.Collections.Generic;
+using System.Linq;
 using Microsoft.PowerFx.Types;
 
 namespace Microsoft.PowerFx.Connectors
 {
     public static class CdpExtensions
     {
-        public static bool TryGetFieldExternalTableName(this RecordType recordType, string fieldName, out string tableName, out string foreignKey)
+        public static bool TryGetFieldRelationships(this RecordType recordType, string fieldName, out IEnumerable<ConnectorRelationship> relationships)
```
Comment (LucGenetier) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/CdpExtensions.cs:12]: >TryGetFieldRelationships [](http://example.com/codeflow?start=27&length=24)

I'll add [Obsolete]

File: src/libraries/Microsoft.PowerFx.Connectors/Tabular/Capabilities/ColumnCapabilities.cs [hunk 13]
```diff
@@ -72,5 +73,32 @@ public static ColumnCapabilities ParseColumnCapabilities(IDictionary<string, IOp
                 IsChoice = isChoice
             });
         }
+
+        public bool Equals(ColumnCapabilities other)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Tabular/Capabilities/ColumnCapabilities.cs:77]: >public bool Equals(ColumnCapabilities other) [](http://example.com/codeflow?start=8&length=44)

Intrinsically,  why would we ever need to compare ColumnCapabilities?

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs (hunk 7 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:23]: A ConnectorType is a sorta converted version of a Swagger schema and here we are comparing all elements we can identify in the schema ... and there are simply a lot
We could imagine converting the schema to a string and make a comparison of that string but the equality won't be valid in all cases as you could define types locally or as a reference, both identical, but get an equality difference (this is just one example)

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorRelationship.cs (hunk 10 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorRelationship.cs:8]: I added [Obsolete]

File: src/libraries/Microsoft.PowerFx.Connectors/Public/CdpExtensions.cs (hunk 12 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/CdpExtensions.cs:12]: This API should not exist but Tahir's team is currently using it here [https://powerbi.visualstudio.com/Insights%20Apps%20Platform/_git/Platform.Service?path=/src/service/FederatedKnowledge/TabularDatasource/Implementation/DatasetTableProviders/TabularDatasetTableProvider.cs&version=GBmain&_a=contents](https://powerbi.visualstudio.com/Insights%20Apps%20Platform/_git/Platform.Service?path=/src/service/FederatedKnowledge/TabularDatasource/Implementation/DatasetTableProviders/TabularDatasetTableProvider.cs&version=GBmain&_a=contents)
We now know they shouldn't be using it (like TryGetUnderlyingFieldType) and this is due to the recent discovery that we cannot reuse Dataverse "syntax" (a.b.c.d ...) for fields having the relationships in the general CDP case (or "shims")

When Greg will have told us which direction to follow, we'll be able to remove all this code (or replace it with an improved and working version for shim-based tabular connectors)

File: src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs [hunk 14]
```diff
@@ -17,6 +17,8 @@
 using Microsoft.PowerFx.Core.Utils;
 using Microsoft.PowerFx.Types;
 using static Microsoft.PowerFx.Connectors.Constants;
+
+#pragma warning disable SA1000 // The keyword 'new' should be followeed by a space
```
[the remark this answers is not in the document]
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs:21]: SA1000 ([https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1000.md](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1000.md)) forces spacing for C# keywords but doesn't consider `Dictionary<string, FormulaValue> dvParams = new();` as valid due to the "missing" space between new and the constructor parenthesis
As I think it's adding some consistency and value I prefer keeping it enabled at the project level and turn it off only in this file
Very debatable, I know...

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.CdpExtensions.TryGetFieldExternalTableName(Microsoft.PowerFx.Types.RecordType,System.String,System.String@,System.String@)

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs [hunk 15]
```diff
@@ -114,12 +114,12 @@ public class ConnectorType : SupportsConnectorErrors
...
-        // Relationships to external tables
-        internal List<string> ExternalTables { get; set; }
+        // Relationships to external tables        
+#pragma warning disable CS0618 // Type or member is obsolete
+        internal IEnumerable<ConnectorRelationship> ExternalTables => _externalTables;
 
-        internal string RelationshipName { get; set; }
-
-        internal string ForeignKey { get; set; }
+        private List<ConnectorRelationship> _externalTables;
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:121]: >List<ConnectorRelationship> [](http://example.com/codeflow?start=16&length=27)

nit: IEnumerable<>
I know its annoying but immutable collections are much easier to maintain/debug.

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs [hunk 16]
```diff
@@ -340,5 +352,132 @@ private Dictionary<string, FormulaValue> GetEnum()
...
+                   left.NotificationUrl == right.NotificationUrl &&
+                   ((left.HiddenRecordType == null && right.HiddenRecordType == null) || left.HiddenRecordType.Equals(right.HiddenRecordType)) &&
+                   left.Binary == right.Binary &&
+                   left.MediaKind == right.MediaKind &&
+                   Enumerable.SequenceEqual((IList<string>)left.ExternalTables ?? Array.Empty<string>(), (IList<string>)right.ExternalTables ?? Array.Empty<string>());
+        }
+
+        internal static bool ArrayEquals(ConnectorType[] leftArray, ConnectorType[] rightArray, int depth)
+        {
+            if (depth > 30)
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:397]: >30 [](http://example.com/codeflow?start=24&length=2)

Can we hoist this out in a constant, Is 30 too big? Ie can we add a test to verify we don't stack overflow with 30 depth.

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs [hunk 17]
```diff
@@ -340,5 +352,132 @@ private Dictionary<string, FormulaValue> GetEnum()
...
+                   left.Visibility == right.Visibility &&
+                   ((left.Capabilities == null && right.Capabilities == null) || left.Capabilities.Equals(right.Capabilities)) &&
+                   left.KeyType == right.KeyType &&
+                   left.KeyOrder == right.KeyOrder &&
+                   left.Permission == right.Permission &&
+                   left.NotificationUrl == right.NotificationUrl &&
+                   ((left.HiddenRecordType == null && right.HiddenRecordType == null) || left.HiddenRecordType.Equals(right.HiddenRecordType)) &&
+                   left.Binary == right.Binary &&
+                   left.MediaKind == right.MediaKind &&
+                   Enumerable.SequenceEqual((IList<string>)left.ExternalTables ?? Array.Empty<string>(), (IList<string>)right.ExternalTables ?? Array.Empty<string>());
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:392]: We should also consider minimalizing it, maybe we should think of this comparison in terms of just coercion? Ie just compare field names... the problem is .Equals doesn't have good documentation on Aggregate type... but from usage it looks like its primarily used for coercion?

Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:392]: Do we have some test, testing equality?

**Review by MikeStall — CHANGES_REQUESTED** [review]
:clock1:

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs (hunk 16 above)
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:397]: We'd hit millions of nodes and blow up by the time we got depth 30.  But we should identify a solution that is fundamentally correct and not approximated by depth level.

File: src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs [hunk 18]
```diff
@@ -340,5 +352,132 @@ private Dictionary<string, FormulaValue> GetEnum()
 
             return enumDisplayNames.Zip(enumValues, (dn, ev) => new KeyValuePair<string, FormulaValue>(dn, ev)).ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
         }
+
+        public bool Equals(ConnectorType other)
+        {
+            return Equals(this, other, 0);                  
+        }
+
+        internal static bool Equals(ConnectorType left, ConnectorType right, int depth)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs:361]: >Equals [](http://example.com/codeflow?start=29&length=6)

This is a "structural type equivalence". 

Is there a stronger sense of identify for the types? (ie, why is name + container not sufficient?)

File: src/libraries/Microsoft.PowerFx.Connectors/Tabular/Capabilities/ColumnCapabilitiesBase.cs [hunk 19]
```diff
@@ -8,5 +8,6 @@ namespace Microsoft.PowerFx.Connectors
     [JsonConverter(typeof(AbstractTypeConverter<ColumnCapabilitiesBase>))]
     internal abstract class ColumnCapabilitiesBase
     {
+        public abstract override int GetHashCode();
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Tabular/Capabilities/ColumnCapabilitiesBase.cs:11]: >public abstract override int GetHashCode();    [](http://example.com/codeflow?start=8&length=46)

we should not need this. 
This is just a poco - the comparison check could be done at a higher semantic level.
